modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj

```diff
@@ -629,35 +631,43 @@
 ;;
 ;; [:time-interval <datetime field> -1 :day]
 ;;
-;;
+(def ^:private temporal-type->arithmetic-function
+  {:timestamp :timestamp_add
+   :datetime  :datetime_add
+   :date      :date_add
+   :time      :time_add})
+
 (defrecord AddIntervalForm [hsql-form amount unit]
   hformat/ToSql
   (to-sql [_]
-    (loop [hsql-form hsql-form]
-      (let [t      (temporal-type hsql-form)
-            add-fn (case t
-                     :timestamp :timestamp_add
-                     :datetime  :datetime_add
-                     :date      :date_add
-                     :time      :time_add
-                     nil)]
-        (if-not add-fn
-          (recur (->temporal-type :datetime hsql-form))
-          (do
-            (assert-addable-unit t unit)
-            (hformat/to-sql (hsql/call add-fn hsql-form (interval amount unit)))))))))
+    (let [t      (temporal-type hsql-form)
+          add-fn (temporal-type->arithmetic-function t)]
+      (hformat/to-sql (hsql/call add-fn hsql-form (interval amount unit))))))
+
+(defn- add-interval-form [hsql-form amount unit]
+  (loop [hsql-form hsql-form]
+    (let [t      (temporal-type hsql-form)
+          add-fn (temporal-type->arithmetic-function t)]
+      (if (or (not add-fn)
+              (and (not (contains? (temporal-type->supported-units t) unit))
+                   (contains? (temporal-type->supported-units :datetime) unit)))
+        (recur (->temporal-type :datetime hsql-form))
```

Comment:
Sure.

1. `(not add-fn)`. Because `(temporal-type hsql-form)` returned `nil` (we don't know the type)
2. `(not (contains? (temporal-type->supported-units t) unit))` because the unit is `:timestamp` and the `unit` is `:month`

Reply:
If you're pointing out that we don't need to use `loop` here since it's not a loop at all then I'm not disagreeing with you. Feel free to change it. I just took the existing code and tweaked it a bit and didn't do cleanup like that.

Reply:
Are 1. and 2. the first and the second iteration? Wouldn't we set `:datetime` after 1. and see it in 2.?
What input for the function do you have in mind?